### PR TITLE
Support buckets with '.' with etcd+coreDNS

### LINKED ISSUE #7340: Unable to delete buckets with a dot(.) in their name through GUI in federated setups
I was able to create a bucket named `he.man` from minio GUI interface. And was also able to delete the bucket from the GUI interface. However, If I refresh the page before deleting the bucket, the bucket name changes to `man/he`. Once this happens we cannot delete the bucket ever from GUI interface and it can only be deleted from the cli.

After page refresh:
image

## Expected Behavior

Either users should not be allowed to create buckets with a dot in the bucket name or there should be a way to delete these buckets from the GUI interface.

There should be a way to delete these buckets from the GUI interface.

## Steps to Reproduce (for bugs)

1. Create a bucket called abc.xyz through minio GUI interface
2.  Refresh the browser page. Now the bucket name changes to xyz/abc
3. Try to delete the bucket through GUI interface. (Throws an error: Bucket name contains invalid characters )

## Your Environment

* Version used (`minio version`): 2019-01-31T00:31:19Z
* Environment name and version (e.g. nginx 1.9.1): NA
* Server type and version: NA
* Operating System and version (`uname -a`): Linux <hostname> 4.4.0-116-generic #140-Ubuntu SMP Mon Feb 12 21:23:04 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
* Link to your project: NA

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
Support buckets with '.' with etcd+coreDNS

## Motivation and Context
Fixes #7340

## Regression
No

## How Has This Been Tested?
Using a local etcd setup

```
~ MINIO_DOMAIN=backspace MINIO_ETCD_ENDPOINTS=http://localhost:2379 minio server ~/test
```

```
~ mc ls --debug myminio-dns/test.bucket
...
mc: <DEBUG> GET /?delimiter=%2F&max-keys=1000&prefix= HTTP/1.1
Host: test.bucket.backspace:9000
...
```
Without this PR we cannot support buckets with `.` in them.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-xl.sh | :x: [more...](#eb740af5-c24f-4da0-b0ae-914ee4fdcbc6) |
| mint-large-bucket.sh | :x: [more...](#e14bbd58-86be-4767-8728-95bf58fe9414) |
| mint-compression-dist-xl.sh | :x: [more...](#02b1ed86-8989-4c9a-889e-4408e6d13d0e) |
<a name="e14bbd58-86be-4767-8728-95bf58fe9414"></a>
### 7353-e088a6e/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.51:32013
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp ca95d7110ed1:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... FAILED in 0 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-yzntwpcjdcnr99",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 304,
  "error": "RequestError: send request failed\ncaused by: Put http://72.28.97.51:32013/aws-sdk-go-test-yzntwpcjdcnr99: dial tcp 72.28.97.51:32013: connect: connection refused",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 14 tests successfully.
```
<a name="eb740af5-c24f-4da0-b0ae-914ee4fdcbc6"></a>
### 7353-e088a6e/mint-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.56:32244
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 6c029cbe0960:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... FAILED in 1 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-wgu9yv5y0k430g",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 273,
  "error": "RequestError: send request failed\ncaused by: Put http://72.28.97.56:32244/aws-sdk-go-test-wgu9yv5y0k430g: dial tcp 72.28.97.56:32244: connect: connection refused",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 14 tests successfully.
```
<a name="02b1ed86-8989-4c9a-889e-4408e6d13d0e"></a>
### 7353-e088a6e/mint-compression-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.57:31638
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp fbe4230145c3:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... FAILED in 6 minutes and 0 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-tcfazc1z3ob91i",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 360013,
  "error": "XMinioServerNotInitialized: Server not initialized, please try again.\n\tstatus code: 503, request id: 158AF06F255C305D, host id: ",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}

Executed 0 out of 14 tests successfully.
```

File: pkg/dns/etcd_dns.go [hunk 1]
```diff
@@ -118,10 +130,8 @@ func (c *coreDNS) list(key string) ([]SrvRecord, error) {
 		// In all other situations when we want to list all DNS records,
 		// which is handled in the else clause.
 		for _, domainName := range c.domainNames {
-			if key != msg.Path(fmt.Sprintf(".%s.", domainName), defaultPrefixPath) {
-				if srvRecord.Key == "/" {
-					srvRecords = append(srvRecords, srvRecord)
-				}
+			if key != "" && key != msg.Path(fmt.Sprintf(".%s.", domainName), defaultPrefixPath) {
+				srvRecords = append(srvRecords, srvRecord)
```
Comment (kannappanr) [inline pkg/dns/etcd_dns.go]: why do you need this `if` check here? `srvRecords = append(srvRecords, srvRecord)` is executed in all scenarios.

Comment (harshavardhana) [reply pkg/dns/etcd_dns.go]: We need to only return back srvRecords which match the key otherwise we will have unexpected entries.

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-compression-dist-xl.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#ecead33b-1a6e-45dd-aa92-e66a80f72827) |
<a name="ecead33b-1a6e-45dd-aa92-e66a80f72827"></a>
### 7353-20a8a6f/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.60:31959
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp ee91bec4a75c:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 1 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 5 minutes and 10 seconds
(4/14) Running aws-sdk-ruby tests ... done in 15 seconds
(5/14) Running awscli tests ... done in 19 minutes and 9 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... FAILED in 1 minutes and 5 seconds
{
  "name": "mc",
  "duration": "30309",
  "function": "test_get_object",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-dir /tmp/.mc-3352 --quiet --no-color cp myminio/mc-test-bucket-12754/mc-test-object-32253 mc-test-object-32253.downloaded  >>> \nmc: <ERROR> Unable to validate source myminio/mc-test-bucket-12754/mc-test-object-32253"
}

Executed 6 out of 14 tests successfully.
```

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :x: [more...](#bad3e66b-9a20-4799-a61d-0c2297540461) |
| mint-compression-dist-xl.sh | :x: [more...](#fed7b509-9c31-42c1-a298-7971df9f9777) |
<a name="bad3e66b-9a20-4799-a61d-0c2297540461"></a>
### 7353-48ec4a9/mint-compression-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.51:30242
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp b885522b03fa:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 0 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 43 seconds
(4/14) Running aws-sdk-ruby tests ... done in 2 seconds
(5/14) Running awscli tests ... done in 1 minutes and 45 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 23 seconds
(8/14) Running minio-dotnet tests ... done in 26 seconds
(9/14) Running minio-go tests ... done in 44 seconds
(10/14) Running minio-java tests ... done in 2 minutes and 2 seconds
(11/14) Running minio-js tests ... done in 37 seconds
(12/14) Running minio-py tests ... FAILED in 4 minutes and 50 seconds
{
  "message": "AccessDenied: message: Access Denied",
  "status": "FAIL",
  "name": "minio-py:test_presigned_get_object_expiry_5sec",
  "function": "presigned_get_object(bucket_name, object_name, expires, response_headers, request_date)",
  "duration": 68,
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1220, in test_presigned_get_object_expiry_5sec\n    object_name).get_exception()\nminio.error.AccessDenied: AccessDenied: message: Access Denied\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1885, in main\n    test_presigned_get_object_expiry_5sec(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 1228, in test_presigned_get_object_expiry_5sec\n    raise Exception(err)\nException: AccessDenied: message: Access Denied\n",
  "args": {
    "bucket_name": "minio-py-test-3b97349e-b0f2-4e42-81e5-ecc57ff140e0",
    "object_name": "94262fd5-000c-4670-b485-c99bb0fffae4"
  }
}

Executed 11 out of 14 tests successfully.
```
<a name="fed7b509-9c31-42c1-a298-7971df9f9777"></a>
### 7353-48ec4a9/mint-compression-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      72.28.97.57:30722
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 7e6dc1e083ea:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 2 minutes and 0 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 45 seconds
(4/14) Running aws-sdk-ruby tests ... done in 5 seconds
(5/14) Running awscli tests ... done in 1 minutes and 50 seconds
(6/14) Running healthcheck tests ... done in 1 seconds
(7/14) Running mc tests ... done in 27 seconds
(8/14) Running minio-dotnet tests ... done in 38 seconds
(9/14) Running minio-go tests ... done in 1 minutes and 2 seconds
(10/14) Running minio-java tests ... done in 4 minutes and 2 seconds
(11/14) Running minio-js tests ... done in 45 seconds
(12/14) Running minio-py tests ... FAILED in 4 minutes and 30 seconds
{
  "name": "minio-py:test_presigned_get_object_expiry_5sec",
  "args": {
    "bucket_name": "minio-py-test-9a0e1254-896d-4117-8a5a-3f8df9afc297",
    "object_name": "ab50d882-7efe-45d7-82c0-5341c9f1bb68"
  },
  "message": "AccessDenied: message: Access Denied",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1220, in test_presigned_get_object_expiry_5sec\n    object_name).get_exception()\nminio.error.AccessDenied: AccessDenied: message: Access Denied\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1885, in main\n    test_presigned_get_object_expiry_5sec(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 1228, in test_presigned_get_object_expiry_5sec\n    raise Exception(err)\nException: AccessDenied: message: Access Denied\n",
  "status": "FAIL",
  "function": "presigned_get_object(bucket_name, object_name, expires, response_headers, request_date)",
  "duration": 165
}

Executed 11 out of 14 tests successfully.
```

**Review by poornas — APPROVED** [review]
Tested, LGTM

**Review by kannappanr — APPROVED** [review]
LGTM Tested.

**Pull request merged**